### PR TITLE
Add the rest of the nodes

### PR DESCRIPTION
**Pull request by davepagurek**
This makes two new base classes:
- `Boolean`, for things in the condition of a `ScopeFn`
- `ScopeFn`, which takes a `NodePtr` in addition to a `Json` in its `to_html` method. It transforms the context and passes it to the `NodePtr`.

Some refactors done in the process:
- Refactored `TagStatement` and `Nested` into one class
- Took the `const` out of the `Json` typedef so that we can use it mutably to build up a new context in a `ScopeFn`. After we're good with all the nodes, we should change the `to_html` signature to take in a `const Json` (we were going to refactor it anyway to add `const` to the method.) Added this to https://github.com/emerald-lang/emerald/issues/78

Things left to do after this:
- I can't `make` the whole thing because there are some other build errors right now, so probably when we fix the existing ones, we'll discover new build errors in this code.
- Our current code for repeated rules will need to be changed, unfortunately. https://github.com/emerald-lang/emerald/issues/80

### CODE REVIEW

File: cpp/src/grammar.cpp [hunk 1]
```diff
@@ -127,15 +120,16 @@ Grammar::Grammar() : emerald_parser(syntax) {
       std::vector<std::string> class_names = repeated<std::string>(sv, 2);
       NodePtr body = present(sv, 3) ? sv[3].get<NodePtr>() : NodePtr();
       NodePtr attributes = present(sv, 4) ? sv[4].get<NodePtr>() : NodePtr();
+      NodePtr nested =
+        present(sv, 5) ? sv[5].get<peg::SemanticValues>()[0].get<NodePtr>() : NodePtr();
 
-      return NodePtr(new TagStatement(tag_name, id_name, class_names, body, attributes));
+      return NodePtr(
+        new TagStatement(tag_name, id_name, class_names, body, attributes, nested));
```
Comment (armcburney) [inline cpp/src/grammar.cpp:127]: Just curious - why did you put this on a new line?

Comment (davepagurek) [reply cpp/src/grammar.cpp:127]: It was getting a little long

Comment (davepagurek) [reply cpp/src/grammar.cpp:127]: I also added another parameter to the list, btw, which is what made me think to bump it down a line

File: cpp/src/grammar.peg [hunk 2]
```diff
@@ -37,7 +35,7 @@ il_lit_str_content     <- !'"' .
 
 escaped                <- '\\' .
 
-tag_statement          <- tag $identifier< id_name? > $classes< class_name* > ~space* $body< text_content? > $attributes< attributes? >
+tag_statement          <- tag id_name? class_name* ~space* text_content? attr_list? (~nl ~lbrace ~nl ROOT ~rbrace ~nl)?
```
Comment (armcburney) [inline cpp/src/grammar.peg:38]: 👍 💯

File: cpp/src/nodes/binary_expr.hpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  enum Operator { AND, OR };
+  static const std::string AND_STR;
+  static const std::string OR_STR;
+
+  BinaryExpr(BooleanPtr lhs, Operator op, BooleanPtr rhs);
+
+  bool truthy(Json &context) const override;
+
+private:
+
```
Comment (armcburney) [inline cpp/src/nodes/binary_expr.hpp:20]: can you remove this blank newline please?

File: cpp/src/nodes/binary_expr.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  BooleanPtr lhs;
+  BooleanPtr rhs;
+  Operator op;
```
Comment (armcburney) [inline cpp/src/nodes/binary_expr.hpp:23]: And add a blank newline after line 23?

File: cpp/src/nodes/each.cpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
+#include "each.hpp"
+#include <vector>
+
+Each::Each(std::string collection_name, std::string val_name, std::string key_name):
+  collection_name(collection_name),
+  val_name(val_name),
+  key_name(key_name)
+{}
+
+std::string Each::to_html(NodePtr body, Json &context) const {
```
Comment (armcburney) [inline cpp/src/nodes/each.cpp:10]: This method is kinda big, and hard to read - do you think we could break this up into a few smaller private methods belonging to the `Each` class?

File: cpp/src/nodes/scope.cpp [hunk 6]
```diff
@@ -1,7 +1,7 @@
 #include "scope.hpp"
 
-Scope::Scope(NodePtr scope_fn, NodePtr body) : scope_fn(scope_fn), body(body) {}
+Scope::Scope(ScopeFnPtr scope_fn, NodePtr body) : scope_fn(scope_fn), body(body) {}
 
 std::string Scope::to_html(Json &context) {
-  return scope_fn.to_html(body);
+  return scope_fn->to_html(body, context);
```
Comment (armcburney) [inline cpp/src/nodes/scope.cpp:6]: Good catch 👍

File: cpp/src/nodes/each.hpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+public:
+  Each(std::string collection_name, std::string val_name, std::string key_name);
+
+  std::string to_html(NodePtr body, Json &context) const override;
+
+private:
+  std::string collection_name;
+  std::string val_name;
+  std::string key_name;
```
Comment (armcburney) [inline cpp/src/nodes/each.hpp:18]: Could we add a blank newline after this statement?

File: cpp/src/nodes/tag_statement.cpp [hunk 8]
```diff
@@ -6,13 +6,15 @@ TagStatement::TagStatement(
   std::string id,
   std::vector<std::string> classes,
   NodePtr body,
-  NodePtr attributes
+  NodePtr attributes,
```
Comment (armcburney) [inline cpp/src/nodes/tag_statement.cpp:9]: So we renamed the `attributes` rule to the `attr_list` rule [here](https://github.com/emerald-lang/emerald/pull/81/files#diff-c33c4b016c1880061bfb23d9f65a525fR130). Should we call this `attr_list` in this class as well as [here](https://github.com/emerald-lang/emerald/pull/81/files#diff-c33c4b016c1880061bfb23d9f65a525fR127) to be consistent?

File: cpp/src/nodes/unary_expr.hpp [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+class UnaryExpr : public Boolean {
+
+public:
+  UnaryExpr(bool negated, BooleanPtr expr);
+
+  bool truthy(Json &context) const override;
+
+private:
+  bool negated;
+  BooleanPtr expr;
```
Comment (armcburney) [inline cpp/src/nodes/unary_expr.hpp:16]: Could we add a newline after here?

File: cpp/src/scopes.peg [hunk 10]
```diff
@@ -1,17 +1,17 @@
...
-each         <- 'each' space* $collection< variable_name > space* 'as' space* $val_name< variable_name > $indexed< (',' space* $key_name< variable_name >)? >
+each         <- ~'each' ~space* variable_name ~space* ~'as' ~space* ~variable_name (~',' ~space* ~variable_name)? >
 
 boolean_expr <- binary_expr / unary_expr
 
-binary_expr  <- $lhs< unary_expr > space+ $op< 'and' / 'or' > space+ $rhs< boolean_expr >
+binary_expr  <- unary_expr ~space+ ('and' / 'or') ~space+ boolean_expr
 
-unary_expr   <- $negated< ('not' space+)? > ($val< variable_name > / '(' space* $val< boolean_expr > space* ')')
+unary_expr   <- ('not' space+)? (variable_name / ~'(' ~space* boolean_expr ~space* ~')')
```
Comment (armcburney) [inline cpp/src/scopes.peg:16]: 👍

File: cpp/src/scopes.peg [hunk 11]
```diff
@@ -1,17 +1,17 @@
...
-with         <- 'with' space* variable_name
+with         <- ~'with' ~space* variable_name
 
-each         <- 'each' space* $collection< variable_name > space* 'as' space* $val_name< variable_name > $indexed< (',' space* $key_name< variable_name >)? >
+each         <- ~'each' ~space* variable_name ~space* ~'as' ~space* ~variable_name (~',' ~space* ~variable_name)? >
 
 boolean_expr <- binary_expr / unary_expr
 
-binary_expr  <- $lhs< unary_expr > space+ $op< 'and' / 'or' > space+ $rhs< boolean_expr >
+binary_expr  <- unary_expr ~space+ ('and' / 'or') ~space+ boolean_expr
```
Comment (armcburney) [inline cpp/src/scopes.peg:14]: 👍

File: cpp/src/scopes.peg [hunk 12]
```diff
@@ -1,17 +1,17 @@
...
+given        <- ~'given' ~space* boolean_expr
 
-unless       <- 'unless' space* boolean_expr
+unless       <- ~'unless' ~space* boolean_expr
 
-with         <- 'with' space* variable_name
+with         <- ~'with' ~space* variable_name
 
-each         <- 'each' space* $collection< variable_name > space* 'as' space* $val_name< variable_name > $indexed< (',' space* $key_name< variable_name >)? >
+each         <- ~'each' ~space* variable_name ~space* ~'as' ~space* ~variable_name (~',' ~space* ~variable_name)? >
```
Comment (armcburney) [inline cpp/src/scopes.peg:10]: 👍

**Review by armcburney — CHANGES_REQUESTED** [review]
Fantastic work 🙌 - just a few minor things, which I commented above.

**Review by armcburney — APPROVED** [review]
We'll start using an autoformatter for C++ once we get the grammar working.

**Review by armcburney — APPROVED** [review]
:100: :shipit:

**Pull request merged**